File: app/src/main/java/com/example/msdc/ui/tv_shows/TvAiringTodayFragment.java
package com.example.msdc.ui.tv_shows;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ProgressBar;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.example.msdc.R;
import com.example.msdc.adapter.TVGridAdapter;
import com.example.msdc.api.ApiClient;
import com.example.msdc.api.ApiService;
import com.example.msdc.api.TVResponse;
import com.example.msdc.api.TVResult;
import com.example.msdc.databinding.FragmentTvAiringTodayBinding;

import java.util.ArrayList;
import java.util.List;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;

public class TvAiringTodayFragment extends Fragment {
    private ApiService apiService;
    private TVGridAdapter tvAiringTodayAdapter;
    private final List<TVResult> tvAiringTodayResults = new ArrayList<>();

    private RecyclerView rvTvAiringToday;
    private ProgressBar loadingTvAiringToday;
    private TextView textNoResult;

    public static final String MY_API_KEY = "9bfd8a12ca22a52a4787b3fd80269ea9";

    public static final String LANGUAGE = "en-US";
    private int page = 1;

    private FragmentTvAiringTodayBinding binding;

    public TvAiringTodayFragment() {
        // Required empty public constructor
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        super.onCreate(savedInstanceState);
        binding = FragmentTvAiringTodayBinding.inflate(inflater, container, false);
        View root = binding.getRoot();

        Retrofit retrofit = ApiClient.getClient();
        apiService = retrofit.create(ApiService.class);

        setOnAiringTV(root);

        return root;
    }

    private void setOnAiringTV(View view) {
        rvTvAiringToday = view.findViewById(R.id.rvTvAiringTodayList);
        loadingTvAiringToday = view.findViewById(R.id.loadingTvAiringTodayList);
        textNoResult = view.findViewById(R.id.textNoTvAiringTodayResult);

        tvAiringTodayAdapter = new TVGridAdapter(tvAiringTodayResults, getContext());
        GridLayoutManager gridLayoutManager = new GridLayoutManager(getContext(),3);

        rvTvAiringToday.setLayoutManager(gridLayoutManager);
        rvTvAiringToday.setAdapter(tvAiringTodayAdapter);
        rvTvAiringToday.setHasFixedSize(true);
        rvTvAiringToday.setItemViewCacheSize(9);
        rvTvAiringToday.setDrawingCacheEnabled(true);
        rvTvAiringToday.setDrawingCacheQuality(View.DRAWING_CACHE_QUALITY_HIGH);
        getOnAiringTV(page);

        rvTvAiringToday.addOnScrollListener(new RecyclerView.OnScrollListener() {
            @Override
            public void onScrolled(@NonNull RecyclerView recyclerView, int dx, int dy) {
                super.onScrolled(recyclerView, dx, dy);

                if(!rvTvAiringToday.canScrollVertically(1)){
                    page++;
                    getOnAiringTV(page);
                }
            }
        });
    }

    private void getOnAiringTV(int page){
        int limit = 9;
        Call<TVResponse> call = apiService.getTvAiringToday(MY_API_KEY, LANGUAGE, page, limit);
        call.enqueue(new Callback<TVResponse>(){
            @Override
            public void onResponse(@NonNull Call<TVResponse> call, @NonNull Response<TVResponse> response) {
                if(response.body() != null){
                    loadingTvAiringToday.setVisibility(View.GONE);
                    rvTvAiringToday.setVisibility(View.VISIBLE);

                    int oldCount = tvAiringTodayResults.size();
                    tvAiringTodayResults.addAll(response.body().getResult());
                    tvAiringTodayAdapter.notifyItemRangeInserted(oldCount, tvAiringTodayResults.size());
                } else if(tvAiringTodayResults.isEmpty()){
                    loadingTvAiringToday.setVisibility(View.GONE);
                    textNoResult.setVisibility(View.VISIBLE);
                }
            }

            @Override
            public void onFailure(@NonNull Call<TVResponse> call, @NonNull Throwable t) {
                loadingTvAiringToday.setVisibility(View.GONE);
                Toast.makeText(getContext(), "Failed To Fetch Airing TV Shows !!!", Toast.LENGTH_SHORT).show();
            }
        });
    }
    @Override
    public void onDestroyView() {
        super.onDestroyView();
        binding = null;
    }
}

File: app/src/main/java/com/example/msdc/ui/tv_shows/TvAiringTodayFragment.kt
package com.example.msdc.ui.tv_shows

import android.content.res.Configuration
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ProgressBar
import android.widget.TextView
import android.widget.Toast
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.example.msdc.adapter.TVVerticalAdapter
import com.example.msdc.api.ApiClient.client
import com.example.msdc.api.ApiService
import com.example.msdc.api.TVResponse
import com.example.msdc.api.TVResult
import com.example.msdc.databinding.FragmentTvAiringTodayBinding
import com.google.firebase.crashlytics.FirebaseCrashlytics
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class TvAiringTodayFragment : Fragment() {
    private var apiService: ApiService? = null

    private var tvAiringTodayAdapter: TVVerticalAdapter? = null

    private val tvAiringTodayResults: MutableList<TVResult> = ArrayList()

    private var page = 1

    private lateinit var loadingTvAiringToday: ProgressBar
    private lateinit var rvTvAiringToday: RecyclerView
    private lateinit var textNoResult: TextView

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        // Inflate the layout for this fragment
        super.onCreate(savedInstanceState)
        FirebaseCrashlytics.getInstance().setCrashlyticsCollectionEnabled(true)
        val binding = FragmentTvAiringTodayBinding.inflate(inflater, container, false)

        loadingTvAiringToday = binding.loadingTvAiringTodayList
        rvTvAiringToday = binding.rvTvAiringTodayList
        textNoResult = binding.textNoTvAiringTodayResult

        val root: View = binding.root
        val retrofit = client

        apiService = retrofit!!.create(ApiService::class.java)
        setOnAiringTV()

        return root
    }

    private fun setOnAiringTV() {
        tvAiringTodayAdapter = TVVerticalAdapter(tvAiringTodayResults, requireContext())

        val mPortraitLayoutManager = GridLayoutManager(context, 3)
        val mLandscapeLayoutManager = GridLayoutManager(context, 5)

        if(resources.configuration.orientation == Configuration.ORIENTATION_LANDSCAPE){
            rvTvAiringToday.layoutManager = mLandscapeLayoutManager
        } else {
            rvTvAiringToday.layoutManager = mPortraitLayoutManager
        }
        rvTvAiringToday.adapter = tvAiringTodayAdapter
        rvTvAiringToday.setHasFixedSize(true)
        rvTvAiringToday.setItemViewCacheSize(9)

        getOnAiringTV(page)

        rvTvAiringToday.addOnScrollListener(object : RecyclerView.OnScrollListener() {
            override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
                super.onScrolled(recyclerView, dx, dy)
                if (!rvTvAiringToday.canScrollVertically(1)) {
                    page++
                    getOnAiringTV(page)
                }
            }
        })
    }

    private fun getOnAiringTV(page: Int) {
        val limit = 15
        val call = apiService?.getTvAiringToday(MY_API_KEY, LANGUAGE, page, limit)

        call!!.enqueue(object : Callback<TVResponse?> {
            override fun onResponse(call: Call<TVResponse?>, response: Response<TVResponse?>) {
                if (response.body() != null) {
                    if (response.body()!!.result!!.isNotEmpty()) {
                        loadingTvAiringToday.visibility = View.GONE
                        rvTvAiringToday.visibility = View.VISIBLE

                        val oldCount = tvAiringTodayResults.size
                        tvAiringTodayResults.addAll(response.body()!!.result!!)
                        tvAiringTodayAdapter?.notifyItemRangeInserted(
                            oldCount,
                            tvAiringTodayResults.size
                        )
                    } else {
                        loadingTvAiringToday.visibility = View.GONE
                        textNoResult.visibility = View.VISIBLE
                    }
                }
            }

            override fun onFailure(call: Call<TVResponse?>, t: Throwable) {
                loadingTvAiringToday.visibility = View.GONE
                Toast.makeText(
                    context, t.message + " cause : " + t.cause,
                    Toast.LENGTH_SHORT
                ).show()
            }
        })
    }

    companion object {
        const val MY_API_KEY = "9bfd8a12ca22a52a4787b3fd80269ea9"
        const val LANGUAGE = "en-US"
    }
}